zarray assignment errors

Assigning an `xarray` to a `zarray` doesn't seem to work:
```cpp
auto a1 = xarray<int>();
auto a2 = xarray<int>({0, 1});
zarray z1(a1);
z1 = a2;
```
```
/home/david/miniforge3/envs/xtensor-zarr-dev/include/zarray/zarray_impl.hpp:462:9: error: no matching function for call to 'xt::xarray_container<xt::uvector<int, std::allocator<int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>::resize(const shape_type&) const'
         m_array.resize(shape);
         ^~~~~~~
```
With `noalias` either:
```cpp
auto a1 = xarray<int>();
auto a2 = xarray<int>({0, 1});
zarray z1(a1);
noalias(z1) = a2;
```
```
/home/david/miniforge3/envs/xtensor-zarr-dev/include/zarray/zassign.hpp:32:17: error: 'const class xt::xarray_container<xt::uvector<int, std::allocator<int> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>' has no member named 'assign_to'; did you mean 'assign'?
             rhs.assign_to(lhs.get_implementation());
             ~~~~^~~~~~~~~
             assign
```
Wrapping `a2` in a `zarray` and assigning it fails at run-time:
```cpp
auto a1 = xarray<int>();
auto a2 = xarray<int>({0, 1});
zarray z1(a1);
zarray z2(a2);
z1 = z2;
```
```
unknown file: Failure
C++ exception with description "callback not found" thrown in the test body.
```

--- a/include/zarray/zarray_impl.hpp
+++ b/include/zarray/zarray_impl.hpp
@@ -418,6 +418,35 @@ namespace xt
      * zarray_wrapper *
      ******************/
 
+    namespace detail
+    {
+        template <class T>
+        struct zarray_wrapper_helper
+        {
+            static inline xarray<T>& get_array(xarray<T>& ar)
+            {
+                return ar;
+            }
+
+            static inline xarray<T>& get_array(const xarray<T>&)
+            {
+                throw std::runtime_error("Cannot return non const array from const array");
+            }
+
+            template <class S>
+            static inline void resize(xarray<T>& ar, S&& shape)
+            {
+                ar.resize(std::forward<S>(shape));
+            }
+
+            template <class S>
+            static inline void resize(const xarray<T>&, S&&)
+            {
+                throw std::runtime_error("Cannot resize const array");
+            }
+        };
+    }
+
     template <class CTE>
     template <class E>
     inline zarray_wrapper<CTE>::zarray_wrapper(E&& e)
@@ -429,7 +458,7 @@ namespace xt
     template <class CTE>
     inline auto zarray_wrapper<CTE>::get_array() -> xarray<value_type>&
     {
-        return m_array;
+        return detail::zarray_wrapper_helper<value_type>::get_array(m_array);
     }
 
     template <class CTE>
@@ -459,13 +488,13 @@ namespace xt
     template <class CTE>
     inline void zarray_wrapper<CTE>::resize(const shape_type& shape)
     {
-        m_array.resize(shape);
+        detail::zarray_wrapper_helper<value_type>::resize(m_array, shape);
     }
 
     template <class CTE>
     inline void zarray_wrapper<CTE>::resize(shape_type&& shape)
     {
-        m_array.resize(std::move(shape));
+        detail::zarray_wrapper_helper<value_type>::resize(m_array, std::move(shape));
     }
 
     template <class CTE>

--- a/include/zarray/zassign.hpp
+++ b/include/zarray/zassign.hpp
@@ -15,6 +15,32 @@
 
 namespace xt
 {
+    namespace detail
+    {
+        template <class Tag>
+        struct zexpression_assigner
+        {
+            // Both E1 and E2 are zarray expressions
+            template <class E1, class E2>
+            static void assign_data(E1& e1, const E2& e2)
+            {
+                e2.assign_to(e1.get_implementation());
+            }
+        };
+
+        template <>
+        struct zexpression_assigner<xtensor_expression_tag>
+        {
+            // E1 is a zarray_expression, E2 is an xtensor_expression
+            template <class E1, class E2>
+            static void assign_data(E1& e1, const E2& e2)
+            {
+                using value_type = typename E2::value_type;
+                e1.template get_array<value_type>() = e2;
+            }
+        };
+    }
+
     template <>
     class xexpression_assigner<zarray_expression_tag>
     {
@@ -29,7 +55,7 @@ namespace xt
             auto shape = uninitialized_shape<dynamic_shape<std::size_t>>(size);
             rhs.broadcast_shape(shape, true);
             lhs.resize(std::move(shape));
-            rhs.assign_to(lhs.get_implementation());
+            detail::zexpression_assigner<xexpression_tag_t<E2>>::assign_data(lhs, rhs);
         }
     };
 

--- a/include/zarray/zdispatcher.hpp
+++ b/include/zarray/zdispatcher.hpp
@@ -203,6 +203,18 @@ namespace xt
         };
 
         template <>
+        struct unary_dispatching_types<xassign_dummy_functor>
+        {
+            using type = zunary_ident_types;
+        };
+
+        template <>
+        struct unary_dispatching_types<xmove_dummy_functor>
+        {
+            using type = zunary_ident_types;
+        };
+
+        template <>
         struct unary_dispatching_types<negate>
         {
             using type = zunary_op_types;

--- a/include/zarray/zdispatching_types.hpp
+++ b/include/zarray/zdispatching_types.hpp
@@ -87,6 +87,8 @@ namespace xt
     template <class T>
     using build_unary_double_t = build_unary_impl_t<T, double>;
 
+    using zunary_ident_types = mpl::transform_t<build_unary_identity_t, z_types>;
+
     using zunary_func_types = detail::concatenate_t<
                                   mpl::transform_t<build_unary_identity_t, z_float_types>,
                                   mpl::transform_t<build_unary_double_t, z_int_types>

--- a/test/test_zarray.cpp
+++ b/test/test_zarray.cpp
@@ -62,6 +62,13 @@ namespace xt
         dc = db;
         EXPECT_EQ(dc.get_array<double>(), b);
         EXPECT_NE(c, b);
+
+        xarray<int> d;
+        xarray<int> e = {0, 1};
+        zarray id(d);
+        zarray ie(e);
+        id = ie;
+        EXPECT_EQ(d, e);
     }
 
     TEST(zarray, move_constructor)
@@ -197,7 +204,7 @@ namespace xt
         EXPECT_EQ(a1, a2);
     }
 
-    /*TEST(zarray, noalias_zarray_assign_xarray)
+    TEST(zarray, noalias_zarray_assign_xarray)
     {
         auto a1 = xarray<int>();
         auto a2 = xarray<int>({0, 1});
@@ -205,14 +212,4 @@ namespace xt
         noalias(z1) = a2;
         EXPECT_EQ(a1, a2);
     }
-
-    TEST(zarray, zarray_assign_zarray)
-    {
-        auto a1 = xarray<int>();
-        auto a2 = xarray<int>({0, 1});
-        zarray z1(a1);
-        zarray z2(a2);
-        z1 = z2;
-        EXPECT_EQ(a1, a2);
-    }*/
 }

--- a/test/test_zarray.cpp
+++ b/test/test_zarray.cpp
@@ -48,7 +48,7 @@ namespace xt
     {
         zdispatcher_t<detail::xassign_dummy_functor, 1>::init();
         zdispatcher_t<detail::xmove_dummy_functor, 1>::init();
-        
+
         xarray<double> a = {{1., 2.}, {3., 4.}};
         xarray<double> b = {{2., 2.}, {3., 4.}};
         xarray<double> c = a;
@@ -90,7 +90,7 @@ namespace xt
     {
         zdispatcher_t<detail::xassign_dummy_functor, 1>::init();
         zdispatcher_t<detail::xmove_dummy_functor, 1>::init();
-        
+
         xarray<double> a = {{1., 2.}, {3., 4.}};
         xarray<double> b = {{2., 2.}, {3., 4.}};
         xarray<double> c = a;
@@ -151,7 +151,7 @@ namespace xt
     {
         xarray<double> a = {{1., 2.}, {3., 4.}};
         zarray za(a);
-        
+
         dynamic_shape<size_t> sh = {3u, 4u, 2u};
         za.resize(sh);
         EXPECT_EQ(a.shape(), sh);
@@ -187,5 +187,32 @@ namespace xt
         shape_type res = za.as_chunked_array().chunk_shape();
         EXPECT_EQ(res, chunk_shape);
     }
-}
 
+    TEST(zarray, zarray_assign_xarray)
+    {
+        auto a1 = xarray<int>();
+        auto a2 = xarray<int>({0, 1});
+        zarray z1(a1);
+        z1 = a2;
+        EXPECT_EQ(a1, a2);
+    }
+
+    TEST(zarray, noalias_zarray_assign_xarray)
+    {
+        auto a1 = xarray<int>();
+        auto a2 = xarray<int>({0, 1});
+        zarray z1(a1);
+        noalias(z1) = a2;
+        EXPECT_EQ(a1, a2);
+    }
+
+    TEST(zarray, zarray_assign_zarray)
+    {
+        auto a1 = xarray<int>();
+        auto a2 = xarray<int>({0, 1});
+        zarray z1(a1);
+        zarray z2(a2);
+        z1 = z2;
+        EXPECT_EQ(a1, a2);
+    }
+}

--- a/test/test_zarray.cpp
+++ b/test/test_zarray.cpp
@@ -197,7 +197,7 @@ namespace xt
         EXPECT_EQ(a1, a2);
     }
 
-    TEST(zarray, noalias_zarray_assign_xarray)
+    /*TEST(zarray, noalias_zarray_assign_xarray)
     {
         auto a1 = xarray<int>();
         auto a2 = xarray<int>({0, 1});
@@ -214,5 +214,5 @@ namespace xt
         zarray z2(a2);
         z1 = z2;
         EXPECT_EQ(a1, a2);
-    }
+    }*/
 }

